pkg/codegen/schema/pulumi.json

```diff
@@ -0,0 +1,509 @@
+            "properties": {
+                "variables": {
+                    "description": "A map from variable name to propertySpec that describes a package's configuration variables.",
+                    "type": "object",
+                    "additionalProperties": {
+                        "$ref": "#/$defs/propertySpec"
+                    }
+                },
+                "required": {
+                    "description": "A list of the names of the package's required configuration variables.",
+                    "type": "array",
+                    "items": {
+                        "type": "string"
+                    }
+                }
+            }
+        },
+        "types": {
+            "description": "A map from type token to complexTypeSpec that describes the set of complex types (i.e. object, enum) defined by this package.",
+            "type": "object",
+            "additionalProperties": {
+                "$ref": "#/$defs/complexTypeSpec"
+            },
+            "propertyNames": {
+                "$ref": "#/$defs/token"
+            }
+        },
+        "provider": {
+            "description": "The provider type for this package.",
+            "$ref": "#/$defs/resourceSpec"
+        },
+        "resources": {
+            "description": "A map from type token to resourceSpec that describes the set of resources and components defined by this package.",
+            "type": "object",
+            "additionalProperties": {
+                "$ref": "#/$defs/resourceSpec"
+            },
+            "propertyNames": {
+                "$ref": "#/$defs/token"
+            }
+        },
+        "functions": {
+            "description": "A map from token to functionSpec that describes the set of functions defined by this package.",
+            "type": "object",
+            "additionalProperties": {
+                "$ref": "#/$defs/functionSpec"
+            },
+            "propertyNames": {
+                "$ref": "#/$defs/token"
+            }
+        },
+        "language": {
+            "description": "Additional language-specific data about the package.",
+            "type": "object"
+        }
+    },
+    "additionalProperties": false,
+    "required": [
+        "name"
+    ],
+    "$defs": {
+        "token": {
+            "title": "Token",
+            "type": "string",
+            "$comment": "In the regex below, the 'module' portion of the token is optional. However, a missing module component creates a '::', which breaks URNs ('::' is the URN delimiter). We have many test schemas that use an empty module component successfully, as they never create URNs; while these are _probably_ the only places that need updating, it might be possible that there are module-less type tokens in the wild elsewhere and we may need to remain compatible with those tokens.",
+            "pattern": "^[^0-9][-a-zA-Z0-9]*:([^0-9][a-zA-Z0-9._/]*)?:[^0-9][a-zA-Z0-9._/]*$"
```

Comment:
This lets patterns in that we don't want. For example: `-:-:-`, `:::::`, `::::`, `.:.:.`.

I don't know if we want a pattern like `foo:ba.:thing.` (note the `.` at the end), but that matches as well.

Reply:
> This lets patterns in that we don't want. For example: -:-:-, :::::, ::::, .:.:..

Yes, this is intentional. The pattern here is not as strict as we necessarily want it to be, but I'm very confident that it will not fail any existing schemas. I intend to tighten this pattern with future changes, but I want to be more certain that we won't accidentally start rejecting schemas that are accepted today.

I think we can improve our confidence here by looking at the set of names the code generators can currently handle properly and then tightening this based on those observations, but I am not going to do so with this PR.

Reply:
(I should note that it feels very very likely that there aren't any packages out here that we'd accidentally reject by tightening some of the aspects you mention, but in general we should be pretty conservative by default)

Reply:
*are, not aren't